Unable to compile with fmt 10.0.0

Thank you for supporting EasyRPG Player.
Before creating an issue, please verify that your bug still occurs in the "continuous builds" of the Player.
You can get them at the bottom of our download section: https://easyrpg.org/player/downloads

---
Please fill in the following fields before submitting an issue:
#### Name of the game:

N/A

#### Player platform:

macOS, Linux

#### Attach files (as a .zip archive or link them)

N/A

#### Describe the issue in detail and how to reproduce it:

Trying to build the project with fmt 10.0.0 fails with the following error:
```
[  7%] Building CXX object CMakeFiles/EasyRPG_Player.dir/src/background.cpp.o
/home/linuxbrew/.linuxbrew/Homebrew/Library/Homebrew/shims/linux/super/g++-11 -DFMT_SHARED -DHAVE_ALSA=1 -DHAVE_FREETYPE=1 -DHAVE_HARFBUZZ=1 -DHAVE_LIBMPG123=1 -DHAVE_LIBSNDFILE=1 -DHAVE_LIBSPEEXDSP=1 -DHAVE_LIBXMP=1 -DHAVE_NATIVE_MIDI=1 -DHAVE_OGGVORBIS=1 -DHAVE_WINE=1 -DNDEBUG -DSUPPORT_AUDIO=1 -DUSE_SDL=2 -DWANT_DRWAV=1 -DWANT_FMMIDI=1 -I/tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src -isystem /home/linuxbrew/.linuxbrew/include/SDL2 -isystem /home/linuxbrew/.linuxbrew/include/pixman-1 -isystem /home/linuxbrew/.linuxbrew/include/freetype2 -isystem /home/linuxbrew/.linuxbrew/include/harfbuzz -isystem /home/linuxbrew/.linuxbrew/include/speex -isystem /home/linuxbrew/.linuxbrew/include/vorbis -O3 -DNDEBUG -std=gnu++14 -fno-math-errno -MD -MT CMakeFiles/EasyRPG_Player.dir/src/background.cpp.o -MF CMakeFiles/EasyRPG_Player.dir/src/background.cpp.o.d -o CMakeFiles/EasyRPG_Player.dir/src/background.cpp.o -c /tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/background.cpp
gmake[2]: *** [CMakeFiles/EasyRPG_Player.dir/build.make:191: CMakeFiles/EasyRPG_Player.dir/src/audio_generic.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
In file included from /tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/string_view.h:23,
                 from /tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/utils.h:29,
                 from /tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/game_battler.h:32,
                 from /tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/autobattle.h:23,
                 from /tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/autobattle.cpp:17:
/home/linuxbrew/.linuxbrew/include/fmt/core.h: In instantiation of ‘constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_value(T&&) [with Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = nonstd::sv_lite::basic_string_view<char>&]’:
/home/linuxbrew/.linuxbrew/include/fmt/core.h:1711:29:   required from ‘constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_arg(T&&) [with bool IS_PACKED = true; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; fmt::v10::detail::type <anonymous> = fmt::v10::detail::type::custom_type; T = nonstd::sv_lite::basic_string_view<char>&; typename std::enable_if<IS_PACKED, int>::type <anonymous> = 0]’
/home/linuxbrew/.linuxbrew/include/fmt/core.h:1827:77:   required from ‘constexpr fmt::v10::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {nonstd::sv_lite::basic_string_view<char, std::char_traits<char> >&, const char* const&}; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; Args = {nonstd::sv_lite::basic_string_view<char, std::char_traits<char> >, const char*}]’
/home/linuxbrew/.linuxbrew/include/fmt/core.h:1844:31:   required from ‘constexpr fmt::v10::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<T>::type>::type ...> fmt::v10::make_format_args(T&& ...) [with Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = {nonstd::sv_lite::basic_string_view<char, std::char_traits<char> >&, const char* const&}]’
/home/linuxbrew/.linuxbrew/include/fmt/core.h:2817:44:   required from ‘std::string fmt::v10::format(fmt::v10::format_string<T ...>, T&& ...) [with T = {nonstd::sv_lite::basic_string_view<char, std::char_traits<char> >&, const char* const&}; std::string = std::__cxx11::basic_string<char>; fmt::v10::format_string<T ...> = fmt::v10::basic_format_string<char, nonstd::sv_lite::basic_string_view<char, std::char_traits<char> >&, const char* const&>]’
/tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/output.h:188:22:   required from ‘void Output::Debug(FmtStr&&, Args&& ...) [with FmtStr = const char (&)[57]; Args = {nonstd::sv_lite::basic_string_view<char, std::char_traits<char> >&, const char* const&}]’
/tmp/easyrpg-player-20230513-22941-1sxcmrz/easyrpg-player-0.8/src/autobattle.cpp:57:17:   required from here
/home/linuxbrew/.linuxbrew/include/fmt/core.h:1691:7: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1691 |       formattable,
      |       ^~~~~~~~~~~
/home/linuxbrew/.linuxbrew/include/fmt/core.h:1691:7: note: ‘formattable’ evaluates to false
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -345,6 +345,7 @@ add_library(${PROJECT_NAME} OBJECT
 	src/state.cpp
 	src/state.h
 	src/std_clock.h
+	src/string_view.cpp
 	src/string_view.h
 	src/system.h
 	src/teleport_target.h

--- a/Makefile.am
+++ b/Makefile.am
@@ -327,6 +327,7 @@ libeasyrpg_player_a_SOURCES = \
 	src/state.cpp \
 	src/state.h \
 	src/std_clock.h \
+	src/string_view.cpp \
 	src/string_view.h \
 	src/system.h \
 	src/teleport_target.h \

--- a/src/game_interpreter.cpp
+++ b/src/game_interpreter.cpp
@@ -2187,7 +2187,7 @@ bool Game_Interpreter::CommandChangeVehicleGraphic(lcf::rpg::EventCommand const&
 	Game_Vehicle* vehicle = Game_Map::GetVehicle(vehicle_id);
 
 	if (!vehicle) {
-		Output::Warning("ChangeVehicleGraphic: Invalid vehicle ID {}", vehicle_id);
+		Output::Warning("ChangeVehicleGraphic: Invalid vehicle ID {}", static_cast<int>(vehicle_id));
 		return true;
 	}
 
@@ -2261,7 +2261,7 @@ bool Game_Interpreter::CommandSetVehicleLocation(lcf::rpg::EventCommand const& c
 			// 0 because we adjust all vehicle IDs by +1 to match the lcf values
 			Output::Debug("SetVehicleLocation: Party referenced");
 		} else {
-			Output::Warning("SetVehicleLocation: Invalid vehicle ID {}", vehicle_id);
+			Output::Warning("SetVehicleLocation: Invalid vehicle ID {}", static_cast<int>(vehicle_id));
 			return true;
 		}
 	}
@@ -3494,7 +3494,7 @@ bool Game_Interpreter::CommandConditionalBranch(lcf::rpg::EventCommand const& co
 		Game_Vehicle* vehicle = Game_Map::GetVehicle(vehicle_id);
 
 		if (!vehicle) {
-			Output::Warning("ConditionalBranch: Invalid vehicle ID {}", vehicle_id);
+			Output::Warning("ConditionalBranch: Invalid vehicle ID {}", static_cast<int>(vehicle_id));
 			return true;
 		}
 

--- a/src/game_interpreter_map.cpp
+++ b/src/game_interpreter_map.cpp
@@ -345,7 +345,7 @@ bool Game_Interpreter_Map::CommandEndShop(lcf::rpg::EventCommand const& /* com *
 
 bool Game_Interpreter_Map::CommandShowInn(lcf::rpg::EventCommand const& com) { // code 10730
 	int inn_type = com.parameters[0];
-	auto inn_price = com.parameters[1];
+	int inn_price = com.parameters[1];
 	// Not used, but left here for documentation purposes
 	// bool has_inn_handlers = com.parameters[2] != 0;
 

--- a/src/output.h
+++ b/src/output.h
@@ -22,16 +22,7 @@
 #include <string>
 #include <iosfwd>
 #include <fmt/core.h>
-#include <lcf/dbstring.h>
-
 #include "filesystem_stream.h"
-
-namespace lcf {
-// FIXME: liblcf doesn't depend on fmt, so we need to add this here to enable fmtlib support for lcf::DBString
-inline fmt::basic_string_view<char> to_string_view(const lcf::DBString& s) {
-	return to_string_view(StringView(s));
-}
-}
 
 enum class LogLevel {
 	Error,

--- a/src/player.cpp
+++ b/src/player.cpp
@@ -1240,7 +1240,7 @@ void Player::SetupBattleTest() {
 		}
 
 		Output::Debug("BattleTest Mode 2k3 troop=({}) background=({}) formation=({}) condition=({}) terrain=({})",
-				args.troop_id, args.background.c_str(), args.formation, args.condition, args.terrain_id);
+				args.troop_id, args.background, static_cast<int>(args.formation), static_cast<int>(args.condition), args.terrain_id);
 	} else {
 		Output::Debug("BattleTest Mode 2k troop=({}) background=({})", args.troop_id, args.background);
 	}

--- a/src/string_view.cpp
+++ b/src/string_view.cpp
@@ -1,0 +1,34 @@
+/*
+ * This file is part of EasyRPG Player.
+ *
+ * EasyRPG Player is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * EasyRPG Player is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with EasyRPG Player. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "string_view.h"
+
+#if FMT_VERSION >= EP_FMT_MODERN_VERSION
+
+#include <fmt/format.h>
+
+auto fmt::formatter<lcf::DBString>::format(const lcf::DBString& s, format_context& ctx) const -> decltype(ctx.out()) {
+	string_view sv(s.data(), s.size());
+	return formatter<string_view>::format(sv, ctx);
+}
+
+auto fmt::formatter<lcf::StringView>::format(const lcf::StringView& s, format_context& ctx) const -> decltype(ctx.out()) {
+	string_view sv(s.data(), s.size());
+	return formatter<string_view>::format(sv, ctx);
+}
+
+#endif

--- a/src/string_view.h
+++ b/src/string_view.h
@@ -22,9 +22,9 @@
 #include <lcf/dbstring.h>
 #include <fmt/core.h>
 
-// FIXME: needed to allow building with fmt 5, older versions are untested.
+// Needed to allow building with fmt 5, older versions are untested.
 #if FMT_VERSION < 60000
-#include <fmt/ostream.h>
+#  include <fmt/ostream.h>
 #endif
 
 using StringView = lcf::StringView;
@@ -33,12 +33,33 @@ using U32StringView = lcf::U32StringView;
 using lcf::ToString;
 using lcf::ToStringView;
 
+// Version required to use the new formatting API
+#define EP_FMT_MODERN_VERSION 80000
+
 // FIXME: liblcf doesn't depend on fmt, so we need to add this here to enable fmtlib support for our StringView.
+#if FMT_VERSION >= EP_FMT_MODERN_VERSION
+template<>
+struct fmt::formatter<lcf::StringView> : fmt::formatter<fmt::string_view> {
+	auto format(const lcf::StringView& s, format_context& ctx) const -> decltype(ctx.out());
+};
+
+template<>
+struct fmt::formatter<lcf::DBString> : formatter<string_view> {
+	auto format(const lcf::DBString& s, format_context& ctx) const -> decltype(ctx.out());
+};
+#else
 namespace nonstd { namespace sv_lite {
 template <typename C, typename T>
 inline fmt::basic_string_view<C> to_string_view(basic_string_view<C,T> s) {
-    return fmt::basic_string_view<C>(s.data(), s.size());
+	return fmt::basic_string_view<C>(s.data(), s.size());
 }
 } }
+
+namespace lcf {
+inline fmt::basic_string_view<char> to_string_view(const lcf::DBString& s) {
+	return to_string_view(StringView(s));
+}
+}
+#endif
 
 #endif

